src/app/directives/popup-close.directive.ts

```diff
@@ -0,0 +1,42 @@
+import { DOCUMENT } from '@angular/common'
+import { Directive, Inject, Output, ElementRef } from '@angular/core'
+import { filter, fromEvent, merge, Observable, switchMapTo, take } from 'rxjs'
+
+@Directive({
+  selector: '[popupClose]'
+})
+export class PopupCloseDirective {
+  @Output()
+  popupClose: Observable<unknown>;
+
+  constructor (
+    @Inject(ElementRef) { nativeElement }: ElementRef<HTMLElement>,
+    @Inject(DOCUMENT) documentRef: Document
+  ) {
+    const esc$: Observable<unknown> = fromEvent<KeyboardEvent>(
+      documentRef,
+      'keydown'
+    ).pipe(filter(({ key }) => {
+      return key === 'Escape'
+    }))
+
+    const clickOutside$ = fromEvent<MouseEvent>(documentRef, 'mousedown').pipe(
+      filter(
+        ({ target }) => {
+          return target instanceof Element && !nativeElement.contains(target)
+        }
+      ),
+      switchMapTo(
+        fromEvent<MouseEvent>(documentRef, 'mouseup').pipe(
```

Comment:
there is also one more Angular-way to do it,  https://angular.io/api/core/HostListener